ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,1273 @@
+            ctru_sys::udsGetNetworkStructApplicationData(
+                &network.network as *const _,
+                appdata_buffer.as_mut_ptr().cast(),
+                appdata_buffer.len(),
+                actual_size.as_mut_ptr(),
+            )
+        })?;
+
+        let actual_size = unsafe { actual_size.assume_init() };
+
+        appdata_buffer.truncate(actual_size);
+        appdata_buffer.shrink_to_fit();
+
+        Ok(appdata_buffer)
+    }
+
+    /// Retrieve app data for the currently connected network.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the service is not connected to a network.
+    /// See [`Uds::connect_network()`] to connect to a network.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// let appdata = uds.get_appdata(None)?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsGetApplicationData")]
+    pub fn get_appdata(&self, max_size: Option<usize>) -> Result<Vec<u8>, Error> {
+        if self.service_status() == ServiceStatus::Disconnected {
+            return Err(Error::NotConnected);
+        }
+
+        let mut appdata_buffer = vec![
+            0u8;
+            max_size
+                .unwrap_or(Self::MAX_APPDATA_SIZE)
+                .min(Self::MAX_APPDATA_SIZE)
+        ];
+
+        let mut actual_size = MaybeUninit::uninit();
+
+        ResultCode(unsafe {
+            ctru_sys::udsGetApplicationData(
+                appdata_buffer.as_mut_ptr().cast(),
+                appdata_buffer.len(),
+                actual_size.as_mut_ptr(),
+            )
+        })?;
+
+        let actual_size = unsafe { actual_size.assume_init() };
+
+        appdata_buffer.truncate(actual_size);
+        appdata_buffer.shrink_to_fit();
+
+        Ok(appdata_buffer)
+    }
+
+    /// Connect to a network.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsConnectNetwork")]
+    pub fn connect_network(
+        &mut self,
+        network: &NetworkScanInfo,
+        passphrase: &[u8],
```

Comment:
Does this always need to be nul-terminated? I Wonder if `&CStr` makes sense here instead of `&[u8]` but not positive what kinds of values might be passed in

Reply:
No, it can be whatever data you like. Games conventionally use a null-terminated string, but it's just raw bytes.